File: scripts/auto-assign.js
#!/usr/bin/env node
// Auto-Coordinator - Automatically assigns next tasks when agents push
// Runs in GitHub Action or locally
// Lead Engineer: 2026-08-25

const fs = require('fs');
const path = require('path');

const TASKS_PATH = path.join(__dirname, '../bridge/tasks.json');
const LOCKS_PATH = path.join(__dirname, '../bridge/locks.json');
const STATE_PATH = path.join(__dirname, '../bridge/state.json');
const AGENTS_PATH = path.join(__dirname, '../bridge/agents.md');

function loadJson(p) {
  try {
    return JSON.parse(fs.readFileSync(p, 'utf8'));
  } catch (e) {
    console.error(`Failed to load ${p}:`, e.message);
    return null;
  }
}

function saveJson(p, data) {
  fs.writeFileSync(p, JSON.stringify(data, null, 2) + '\n', 'utf8');
}

function log(msg) {
  console.log(`[AUTO-COORDINATOR ${new Date().toISOString()}] ${msg}`);
}

function main() {
  log('Starting auto-assign check...');

  const tasksData = loadJson(TASKS_PATH);
  const locksData = loadJson(LOCKS_PATH);
  const stateData = loadJson(STATE_PATH);

  if (!tasksData || !locksData || !stateData) {
    log('Missing bridge files, aborting');
    process.exit(1);
  }

  let changed = false;
  const newlyAssigned = [];

  // 1. Check for completed tasks - release their locks if still present but task is done
  const doneTaskIds = tasksData.tasks.filter(t => t.status === 'done').map(t => t.id);
  const activeLocks = locksData.locks || [];
  const locksToKeep = [];

  for (const lock of activeLocks) {
    if (doneTaskIds.includes(lock.taskId)) {
      log(`Releasing lock for done task ${lock.taskId} (agent ${lock.agent})`);
      changed = true;
      // Don't keep this lock
    } else {
      locksToKeep.push(lock);
    }
  }

  if (locksToKeep.length !== activeLocks.length) {
    locksData.locks = locksToKeep;
    locksData.lastUpdate = new Date().toISOString();
    saveJson(LOCKS_PATH, locksData);
    changed = true;
  }

  // 2. Auto-assign next tasks where dependencies are met
  const taskMap = {};
  tasksData.tasks.forEach(t => taskMap[t.id] = t);

  for (const task of tasksData.tasks) {
    if (task.status !== 'todo') continue;

    // Check dependencies
    const deps = task.dependencies || [];
    const depsMet = deps.every(depId => {
      const dep = taskMap[depId];
      return dep && dep.status === 'done';
    });

    if (!depsMet) {
      log(`Task ${task.id} waiting for dependencies: ${deps.filter(d => taskMap[d]?.status !== 'done').join(', ')}`);
      continue;
    }

    // Check file lock conflicts
    const claimedFiles = task.claimedFiles || [];
    const conflict = locksData.locks.some(lock => {
      return lock.files.some(lockedFile => {
        return claimedFiles.some(cf => {
          // Simple glob check: if cf contains * or lockedFile contains *
          if (cf.includes('*') || lockedFile.includes('*')) {
            const base1 = cf.split('*')[0];
            const base2 = lockedFile.split('*')[0];
            return base1 && base2 && (lockedFile.startsWith(base1) || cf.startsWith(base2));
          }
          return cf === lockedFile;
        });
      });
    });

    if (conflict) {
      log(`Task ${task.id} has file lock conflict, skipping`);
      continue;
    }

    // Auto-assign!
    log(`Auto-assigning task ${task.id} to ${task.assignee} (role: ${task.role})`);
    task.status = 'in_progress';
    task.claimedAt = new Date().toISOString();
    newlyAssigned.push(task);

    // Add lock
    locksData.locks.push({
      agent: task.assignee,
      files: claimedFiles,
      since: new Date().toISOString(),
      taskId: task.id
    });

    changed = true;
  }

  if (newlyAssigned.length > 0) {
    tasksData.lastUpdate = new Date().toISOString();
    saveJson(TASKS_PATH, tasksData);
    locksData.lastUpdate = new Date().toISOString();
    saveJson(LOCKS_PATH, locksData);

    // Update state.json
    stateData.lastUpdate = new Date().toISOString();
    stateData.currentTasks = {};
    tasksData.tasks.forEach(t => {
      stateData.currentTasks[t.id] = t.status;
    });
    stateData.lastCoordinatorMessage = `Auto-assigned ${newlyAssigned.length} tasks: ${newlyAssigned.map(t => t.id).join(', ')} at ${new Date().toISOString()}`;
    if (stateData.inProgress) {
      stateData.inProgress = stateData.inProgress.filter(i => !i.includes('Auto-assigned'));
      stateData.inProgress.push(`Auto-assigned ${newlyAssigned.length} tasks: ${newlyAssigned.map(t => t.id).join(', ')}`);
    }
    saveJson(STATE_PATH, stateData);

    log(`Successfully auto-assigned ${newlyAssigned.length} tasks`);
    
    // Output for GitHub Action
    if (process.env.GITHUB_OUTPUT) {
      fs.appendFileSync(process.env.GITHUB_OUTPUT, `assigned=${newlyAssigned.map(t => t.id).join(',')}\n`);
      fs.appendFileSync(process.env.GITHUB_OUTPUT, `count=${newlyAssigned.length}\n`);
    }
  } else {
    log('No new tasks to auto-assign');
  }

  // 3. Summary
  log('=== CURRENT STATUS ===');
  tasksData.tasks.forEach(t => {
    log(`${t.priority} ${t.id}: ${t.status} (${t.assignee}) deps: ${(t.dependencies||[]).join(',')||'none'}`);
  });
  log(`Locks: ${locksData.locks.length} active`);
  locksData.locks.forEach(l => log(`  - ${l.agent} -> ${l.taskId}: ${l.files.join(', ')}`));

  if (changed) {
    log('Changes made - should commit & push');
    // For GitHub Action, set output
    if (process.env.GITHUB_OUTPUT) {
      fs.appendFileSync(process.env.GITHUB_OUTPUT, `changed=true\n`);
    }
    process.exit(0); // Signal that changes were made
  } else {
    log('No changes needed');
    if (process.env.GITHUB_OUTPUT) {
      fs.appendFileSync(process.env.GITHUB_OUTPUT, `changed=false\n`);
    }
  }
}

main();

File: scripts/auto-assign.py
#!/usr/bin/env python3
"""
Auto-Coordinator - Python version
Automatically assigns next tasks when agents push
Compatible with JS version - same logic
"""

import json
import os
from datetime import datetime, timezone
from pathlib import Path

REPO_PATH = Path(__file__).parent.parent
TASKS_PATH = REPO_PATH / 'bridge/tasks.json'
LOCKS_PATH = REPO_PATH / 'bridge/locks.json'
STATE_PATH = REPO_PATH / 'bridge/state.json'

def now():
    return datetime.now(timezone.utc).isoformat()

def log(msg):
    print(f"[AUTO-COORDINATOR {now()}] {msg}")

def load_json(p):
    try:
        with open(p, 'r') as f:
            return json.load(f)
    except Exception as e:
        log(f"Failed to load {p}: {e}")
        return None

def save_json(p, data):
    with open(p, 'w') as f:
        json.dump(data, f, indent=2, ensure_ascii=False)
        f.write('\n')

def main():
    log("Starting auto-assign check (Python)...")
    
    tasks_data = load_json(TASKS_PATH)
    locks_data = load_json(LOCKS_PATH)
    state_data = load_json(STATE_PATH)
    
    if not all([tasks_data, locks_data, state_data]):
        log("Missing bridge files, aborting")
        return False
    
    changed = False
    newly_assigned = []
    
    # 1. Release locks for done tasks
    done_ids = [t['id'] for t in tasks_data['tasks'] if t['status'] == 'done']
    active_locks = locks_data.get('locks', [])
    kept_locks = []
    
    for lock in active_locks:
        if lock['taskId'] in done_ids:
            log(f"Releasing lock for done task {lock['taskId']} (agent {lock['agent']})")
            changed = True
        else:
            kept_locks.append(lock)
    
    if len(kept_locks) != len(active_locks):
        locks_data['locks'] = kept_locks
        locks_data['lastUpdate'] = now()
        save_json(LOCKS_PATH, locks_data)
        changed = True
    
    # 2. Auto-assign next tasks where deps met
    task_map = {t['id']: t for t in tasks_data['tasks']}
    
    for task in tasks_data['tasks']:
        if task['status'] != 'todo':
            continue
        
        # Check dependencies
        deps = task.get('dependencies', [])
        deps_met = all(task_map.get(dep) and task_map[dep]['status'] == 'done' for dep in deps)
        
        if not deps_met:
            waiting = [d for d in deps if not task_map.get(d) or task_map[d]['status'] != 'done']
            log(f"Task {task['id']} waiting for: {', '.join(waiting)}")
            continue
        
        # Check file conflicts
        claimed = task.get('claimedFiles', [])
        conflict = False
        for lock in locks_data['locks']:
            for lf in lock['files']:
                for cf in claimed:
                    # Simple glob check
                    if '*' in cf or '*' in lf:
                        base1 = cf.split('*')[0]
                        base2 = lf.split('*')[0]
                        if base1 and base2 and (lf.startswith(base1) or cf.startswith(base2)):
                            conflict = True
                            break
                    elif cf == lf:
                        conflict = True
                        break
                if conflict:
                    break
            if conflict:
                break
        
        if conflict:
            log(f"Task {task['id']} has file lock conflict, skipping")
            continue
        
        # Auto-assign!
        log(f"Auto-assigning {task['id']} to {task['assignee']} (role: {task['role']})")
        task['status'] = 'in_progress'
        task['claimedAt'] = now()
        newly_assigned.append(task)
        
        locks_data['locks'].append({
            'agent': task['assignee'],
            'files': claimed,
            'since': now(),
            'taskId': task['id']
        })
        
        changed = True
    
    if newly_assigned:
        tasks_data['lastUpdate'] = now()
        save_json(TASKS_PATH, tasks_data)
        locks_data['lastUpdate'] = now()
        save_json(LOCKS_PATH, locks_data)
        
        # Update state.json
        state_data['lastUpdate'] = now()
        state_data['currentTasks'] = {t['id']: t['status'] for t in tasks_data['tasks']}
        state_data['lastCoordinatorMessage'] = f"Auto-assigned {len(newly_assigned)} tasks (Python): {', '.join(t['id'] for t in newly_assigned)} at {now()}"
        if 'inProgress' in state_data:
            state_data['inProgress'] = [i for i in state_data['inProgress'] if 'Auto-assigned' not in i]
            state_data['inProgress'].append(f"Auto-assigned {len(newly_assigned)} tasks (Python): {', '.join(t['id'] for t in newly_assigned)}")
        save_json(STATE_PATH, state_data)
        
        log(f"Successfully auto-assigned {len(newly_assigned)} tasks (Python)")
        
        # GitHub Actions output
        if os.environ.get('GITHUB_OUTPUT'):
            with open(os.environ['GITHUB_OUTPUT'], 'a') as f:
                f.write(f"assigned={','.join(t['id'] for t in newly_assigned)}\n")
                f.write(f"count={len(newly_assigned)}\n")
                f.write("changed=true\n")
    else:
        log("No new tasks to auto-assign")
        if os.environ.get('GITHUB_OUTPUT'):
            with open(os.environ['GITHUB_OUTPUT'], 'a') as f:
                f.write("changed=false\n")
    
    # Summary
    log("=== CURRENT STATUS (Python) ===")
    for t in tasks_data['tasks']:
        log(f"{t['priority']} {t['id']}: {t['status']} ({t['assignee']}) deps:{','.join(t.get('dependencies',[])) or 'none'}")
    log(f"Locks: {len(locks_data['locks'])} active")
    for l in locks_data['locks']:
        log(f"  - {l['agent']} -> {l['taskId']}: {', '.join(l['files'][:2])}...")
    
    return changed

if __name__ == '__main__':
    changed = main()
    if changed:
        log("Changes made - should commit & push")
    else:
        log("No changes needed")
